aiven/resource_flink_table.go

```diff
@@ -0,0 +1,219 @@
+		return diag.Errorf("error setting Flink Tables `table_id` for resource %s: %s", d.Id(), err)
+	}
+	if err := d.Set("table_name", r.TableName); err != nil {
+		return diag.Errorf("error setting Flink Tables `table_name` for resource %s: %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func resourceFlinkTableState(ctx context.Context, d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
+	di := resourceFlinkTableRead(ctx, d, m)
+	if di.HasError() {
+		return nil, fmt.Errorf("cannot get flink table: %v", di)
+	}
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func resourceFlinkTableCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*aiven.Client)
+
+	project := d.Get("project").(string)
+	serviceName := d.Get("service_name").(string)
+	integrationId := d.Get("integration_id").(string)
+	jdbcTable := d.Get("jdbc_table").(string)
+	kafkaTopic := d.Get("kafka_topic").(string)
+	likeOptions := d.Get("like_options").(string)
+	tableName := d.Get("table_name").(string)
+	partitionedBy := d.Get("partitioned_by").(string)
+	schemaSQL := d.Get("schema_sql").(string)
+
+	createRequest := aiven.CreateFlinkTableRequest{
+		IntegrationId: integrationId,
+		JDBCTable:     jdbcTable,
+		KafkaTopic:    kafkaTopic,
+		LikeOptions:   likeOptions,
+		Name:          tableName,
+		PartitionedBy: partitionedBy,
+		SchemaSQL:     schemaSQL,
+	}
+
+	conf := &resource.StateChangeConf{
+		Pending: []string{"CREATING"},
+		Target:  []string{"CREATED"},
+		Refresh: func() (interface{}, string, error) {
+			r, err := client.FlinkTables.Create(project, serviceName, createRequest)
+			if err != nil {
+				if aiven.IsNotFound(err) {
+					// if the integration could not be found we want to retry
+					return nil, "CREATING", nil
+				}
+				return nil, "", err
+			}
+			return r, "CREATED", nil
+		},
+		Delay:      1 * time.Second,
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		MinTimeout: 1 * time.Minute,
+	}
+
+	r, err := conf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tableId := r.(*aiven.CreateFlinkTableResponse).TableId
+	if err := d.Set("table_id", tableId); err != nil {
+		return diag.Errorf("error setting Flink Tables `table_id` for resource %s: %s", d.Id(), err)
+	}
+	d.SetId(buildResourceID(project, serviceName, tableId))
+
+	return resourceFlinkTableRead(ctx, d, m)
+}
+
+func resourceFlinkTableDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*aiven.Client)
+
+	project := d.Get("project").(string)
+	serviceName := d.Get("service_name").(string)
+	tableId := d.Get("table_id").(string)
+
+	conf := &resource.StateChangeConf{
+		Pending: []string{"REMOVING"},
+		Target:  []string{"DELETED"},
+		Refresh: func() (interface{}, string, error) {
+			err := client.FlinkTables.Delete(
+				project,
+				serviceName,
+				aiven.DeleteFlinkTableRequest{
+					TableId: tableId,
+				})
+			if err != nil {
```

Comment:
It looks like we ignore all the errors here. What if Aiven has downtime, and we get a random 5XX error? And in this case, we will continue deleting until the timeout is reached. I would rather ignore 404 but expose all other cases to the user.

Reply:
we can drop the waiter i think and only ignore 404